crates/ruff/src/rules/ruff/rules/ambiguous_unicode_character.rs

```diff
@@ -103,50 +101,133 @@ pub(crate) fn ambiguous_unicode_character(
+        } else if let Some(representant) = CONFUSABLES.get(&(current_char as u32)).copied() {
+            // The current word contains an ambiguous unicode character.
+            if let Some(diagnostic) = diagnostic_for_char(
+                current_char,
+                representant as char,
+                TextSize::try_from(relative_offset).unwrap() + range.start(),
+                context,
+                settings,
+            ) {
+                buffer.push(diagnostic);
+            }
+        } else {
+            // The current word contains at least one unambiguous unicode character.
+            flags |= WordFlags::UNAMBIGUOUS_UNICODE;
         }
     }
 
+    // End of the text.
+    if !buffer.is_empty() {
+        if flags.is_candidate_word() {
+            diagnostics.append(&mut buffer);
+        }
+        buffer.clear();
+    }
+
     diagnostics
 }
+
+bitflags! {
+    #[derive(Default, Debug, Copy, Clone, PartialEq, Eq)]
+    pub struct WordFlags: u8 {
+        /// The word contains at least one ASCII character (like `B`).
+        const ASCII = 0b0000_0001;
+        /// The word contains at least one unambiguous unicode character (like `β`).
+        const UNAMBIGUOUS_UNICODE = 0b0000_0010;
+    }
+}
+
+impl WordFlags {
+    /// Return `true` if the flags indicate that the word is a candidate for flagging
+    /// ambiguous unicode characters.
+    ///
+    /// We follow VS Code's logic for determining whether ambiguous unicode characters within a
+    /// given word should be flagged, i.e., we flag a word if it contains at least one ASCII
+    /// character, or is purely unicode but _only_ consists of ambiguous characters.
+    const fn is_candidate_word(self) -> bool {
+        self.contains(WordFlags::ASCII) || !self.contains(WordFlags::UNAMBIGUOUS_UNICODE)
+    }
+}
+
+/// Create a [`Diagnostic`] to report an ambiguous unicode character.
+fn diagnostic_for_char(
+    confusable: char,
+    representant: char,
+    offset: TextSize,
+    context: Context,
+    settings: &Settings,
+) -> Option<Diagnostic> {
+    if !settings.allowed_confusables.contains(&confusable) {
+        let char_range = TextRange::at(offset, confusable.text_len());
+        let mut diagnostic = Diagnostic::new::<DiagnosticKind>(
+            match context {
+                Context::String => AmbiguousUnicodeCharacterString {
+                    confusable,
+                    representant,
+                }
+                .into(),
+                Context::Docstring => AmbiguousUnicodeCharacterDocstring {
+                    confusable,
+                    representant,
+                }
+                .into(),
+                Context::Comment => AmbiguousUnicodeCharacterComment {
+                    confusable,
+                    representant,
+                }
+                .into(),
+            },
+            char_range,
+        );
+        if settings.rules.enabled(diagnostic.kind.rule()) {
```

Comment:
Unrelated to this PR: I just noticed this now. Our current approach of creating the diagnostic first and then filtering it out is kind of expensive because it means that we format (and allocate) all messages to then immediately drop them again.